modules/gateway/peers.go

```diff
@@ -205,10 +201,19 @@ func (g *Gateway) managedAcceptConnNewPeer(conn net.Conn, remoteVersion string)
 		},
 		sess: muxado.Server(conn),
 	})
-	if pingSucceeded {
-		g.addNode(remoteAddr)
-		return g.save()
-	}
+
+	// Attempt to ping the supplied address. If successful, we will add
+	// remoteAddr to our node list after accepting the peer.
+	go func() {
+		err := g.pingNode(remoteAddr)
+		if err == nil {
+			g.mu.Lock()
+			g.addNode(remoteAddr)
+			g.save()
+			g.mu.Unlock()
+		}
+	}()
```

Comment:
Can you extend the comments here to explain why this needs to be done synchronously instead of asynchronously? I believe you that there's a bug but it should be explained so we don't switch it back in the future.

Ideally there'd be a test to run with it as well, but it seems like the problem doesn't really happen on localhost.